Improve docs (#265)
rubocop/rubocop-ast: docs/modules/ROOT/pages/node_pattern.adoc

end
----

== Ruby Abstract Syntax Tree (AST)

Parser translates Ruby source code to a tree structure represented in text.
A simple integer literal like `1` is represented by `(int 1)` in the AST.
A method call with two integer literals:

[source,ruby]
----
foo(1, 2)
----

is represented with:

[source]
----
(send nil :foo
  (int 1)
  (int 2)
)
----

> Every node is represented with a sequence of children, the first of which is the node type, and others are optionally present and depend on the node type and the parsed source Ruby code.

Reviewer: Given `#children` does not return the type, I think this should be tweaked. Elements in parenthesis are really the type and then children.

Author: Updated.

[source,ruby]
----
> puts RuboCop::AST::ProcessedSource.new('foo(1)', RUBY_VERSION.to_f).ast.to_s
(send nil :foo
  (int 1))
----

== Basic Node Pattern Structure

The simplest Node Pattern would match just the node type.
E.g. the `int` node pattern would match the `(int 1)` AST (literal `1` in Ruby code).
More sophisticated node patterns match more than one child.

== `(` and `)` to Match Children

Several matchers surrounded by parentheses would match a node with children each matching a corresponding matcher, order-dependently.
Ruby code with an array with two integer literals, `[1, 2]` represented in AST as `(array (int 1) (int 2))` could be matched with `(array int int)` node pattern.

For a literal integer, e.g. `1` Ruby code represented by `(int 1)` in AST:

* `int` node pattern will match exactly the node, looking only the node type
* `(int 1)` node pattern will match precisely the node
* `(int 2)` node pattern will not match

== `(` and `)` for Nested Matching

Ruby code with an method call with two integer literals as arguments, `foo(1, 2)` represented in AST as `(send nil :foo (int 1) (int 2))` could be matched with `(send nil? :foo int int)` node pattern.

> To match just those method calls where the first argument is a literal `1`, use `(send nil? :foo (int 1) int`.

Reviewer: ```suggestion
To match just those method calls where the first argument is a literal `1`, use `(send nil? :foo (int 1) int)`.
```

> Any child that is a node, can be a target for nested matching.

Reviewer: ```suggestion
Any child that is a node can be a target for nested matching.
```

end
----

== Ruby Abstract Syntax Tree (AST)

Parser translates Ruby source code to a tree structure represented in text.
A simple integer literal like `1` is represented by `(int 1)` in the AST.
A method call with two integer literals:

[source,ruby]
----
foo(1, 2)
----

is represented with:

[source]
----
(send nil :foo
  (int 1)
  (int 2)
)
----

Every node is represented with a sequence.
The first element is the node type.

> Other elements are child nodes. They are optionally present and depend on the node type.

Reviewer: Since children are not necessarily nodes, best to write:
```suggestion
Other elements are the children. They are optionally present and depend on the node type.
```

`(int [odd? positive?])` - is an int and the value should be odd and positive.

NOTE: Refer to <<Predicate methods>> to see how `odd?` works.

== `!` for Negation

Node pattern `(send nil? :sum !int _)` would match a `sum` call where the first argument is *not* a literal integer.
E.g.:

* it will match `sum(2.0, 3)`, as the first argument is of a `float` type
* it will not match `sum(2, 3)`, as the first argument is of an `int` type

> NOTE: Negation operator works with other node pattern syntax elements, `{}`, `[]`, `()`, `$`, but not with `<>`.

Author: I'm uncertain if this is intentional.
`(send nil? :foo <int ...>)` would work to match a `foo` call with at least one integer literal.
`(send nil? :foo !<int ...>)` would not work to match a `foo` call with no literal integer arguments. Is there a more canonical way to achieve that?

Reviewer: Actually it's more complex than that. You can negate a pattern that targets a single element (e.g. `!{int | sym}`) but not other pattern (e.g. `!{int int | sym sym}`).
Not sure why you wrote it doesn't work with `$`

Author: Got it.
I understand that with variable number of elements like `{int | sym sym}` and `<int ...>` it's an easily justifiable behaviour.

I guess !<true false>` is more tricky due to the hidden number of elements ambiguity? It seems like it would match if there are no arguments (`foo()`), a single non-boolean argument (`foo(1)`), and multiple arguments as well, which just would open space for more confusion.

> Not sure why you wrote it doesn't work with `$`

Because I didn't! 😄

Thanks for the clarification! I've updated this section.
